packages/modules/vehicles/bmw/api.py

```diff
@@ -41,19 +46,63 @@ def dump_json(data: dict, fout: str):
         os.system("sudo chmod 0666 " + replyFile)
 
 
-# ---------------Helper Function-------------------------------------------
 def get_random_string(length: int) -> str:
     letters = string.ascii_letters
     result_str = ''.join(random.choice(letters) for i in range(length))
     return result_str
 
 
-def create_auth_string(client_id: str, client_password: str) -> str:
-    auth_string = client_id + ':' + client_password
-    auth_bytes = auth_string.encode("ascii")
-    b64bytes = base64.b64encode(auth_bytes)
-    auth_string = 'Basic ' + b64bytes.decode("ascii")
-    return auth_string
+def create_s256_code_challenge(code_verifier: str) -> str:
+    """Create S256 code_challenge with the given code_verifier."""
+    data = hashlib.sha256(code_verifier.encode("ascii")).digest()
+    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("UTF-8")
+
+
+# initialize store structures when no store is available
+def init_store():
+    global store
+    store = {}
+    store['Token'] = {}
+    store['expires_at'] = int(0)
+
+
+# load store from file, initialize store structure if no file exists
+def load_store():
+    global store
+    global storeFile
+    try:
+        tf = open(storeFile, 'r', encoding='utf-8')
+        store = json.load(tf)
+        if 'Token' not in store:
+            init_store()
+        tf.close()
+    except FileNotFoundError:
+        log.warning("load_store: store file not found, new authentication required")
+        store = {}
+        init_store()
+    except Exception as e:
+        log.error("init: loading stored data failed, file: " +
+                  storeFile + ", error=" + str(e))
+        store = {}
+        init_store()
+
+
+# write store file
+def write_store():
+    global store
+    global storeFile
+    try:
+        tf = open(storeFile, 'w', encoding='utf-8')
+    except Exception as e:
+        log.error("write_store_file: Exception " + str(e))
+        os.system("sudo rm -f " + storeFile)
+        tf = open(storeFile, 'w', encoding='utf-8')
+    json.dump(store, tf, indent=4)
+    tf.close()
+    try:
+        os.chmod(storeFile, 0o777)
+    except Exception as e:
+        os.system("sudo chmod 0777 " + storeFile + ', error=' + str(e))
```

Comment:
Bitte keine Ausführungsrechte für eine Datendatei setzen.
```suggestion
        os.chmod(storeFile, 0o666)
    except Exception as e:
        os.system("sudo chmod 0666 " + storeFile + ', error=' + str(e))
```

Reply:
ist geändert.